client.js

```diff
@@ -0,0 +1,254 @@
+'use strict';
+
+const { MTProto } = require('@mtproto/core');
+const {
+  addUsersToMessages,
+  getFormedMessagesFromUpdates,
+} = require('./helpers');
+
+module.exports = class TelegramClient {
+  constructor({ apiId, apiHash, test }) {
+    this.apiID = apiId;
+    this.apiHash = apiHash;
+    this.test = test;
+    this.channelMessageSubscribers = {};
+  }
+
+  async connect(authKeys) {
+    if (this.client) return;
+
+    return await new Promise((resolve, reject) => {
+      this.client = new MTProto({
+        api_id: this.apiID,
+        api_hash: this.apiHash,
+        test: this.test,
+        authKeys,
+      });
+
+      const socket = Object.values(this.client.rpcs)[0].transport.socket;
+
+      const onOpen = () => {
+        this.client.updates.on('updates', ({ updates, users }) => {
+          const messages = getFormedMessagesFromUpdates({ updates, users });
+          const messagesByChannelId = {};
+          Object.values(messages).forEach(message => {
+            const channelId = message.channelId;
+            messagesByChannelId[channelId] = {
+              ...messagesByChannelId[channelId],
+              [message.id]: message,
+            };
+          });
+          for (const channelId in messagesByChannelId) {
+            if (this.channelMessageSubscribers[channelId])
+              this.channelMessageSubscribers[channelId](messagesByChannelId[channelId]);
+          }
+        });
+
+        resolve();
+      };
+
+      const onError = (error) => {
+        reject(error);
+      };
+
+      // Node socket uses different methods than browser socket
+      if (socket.on) {
+        socket.on('connect', onOpen);
+        socket.on('error', onError);
+      } else {
+        socket.onopen = onOpen;
+        socket.onerror = onError;
+      }
+    });
+  }
+
+  disconnect() {
+    if (!this.client) return;
+
+    Object.values(this.client.rpcs).forEach(({ transport }) => {
+      transport.emit('close');
+    });
+
+    delete this.client;
+  }
+
+  isConnected() {
+    return !!this.client;
+  }
+
+  importAuthKeys(authKeys) {
+    this.client.importAuthKeys(authKeys);
+  }
+
+  getAuthKeys() {
+    return this.client.getAuthKeys();
+  }
+
+  subscribeToChannelMessages(channelId, callback) {
+    this.channelMessageSubscribers[channelId] = callback;
+  }
+
+  unsubscribeFromChannelMessages(channelId) {
+    delete this.channelMessageSubscribers[channelId];
+  }
+
+  async request(...args) {
+    return await this.client
+      .call(...args)
+      .catch(async error => {
+        if (error.error_message.includes('_MIGRATE_')) {
```

Comment:
since when do errors have `.error_message` instead of `.message`. I fear this error handling code will itself error when given a standard instance of Error. You might consider adding something like:

```js
const errorMessage = error.error_message || error.message || `${error}`
```

Reply:
nice good call